### PR TITLE
feat(smooth): Adds the smooth() operator, the cool operator

### PR DESCRIPTION
**Pull request by jayphelps**
Demo: http://jsbin.com/wufawuq/edit?js,output (turn your volume up)
Prior art: https://en.wikipedia.org/wiki/Smooth_Operator

This is mostly a port of this video to RxJS https://www.youtube.com/watch?v=4TYv2PhG89A

Credit: I can't remember who first came up with this. May have been @benlesh. Let's just say it was all of us.

### CODE REVIEW

File: spec/operators/smooth-spec.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import marbleTestingSignature = require('../helpers/marble-testing'); // tslint:disable-line:no-require-imports
+
+declare const { asDiagram };
+declare const hot: typeof marbleTestingSignature.hot;
+declare const cold: typeof marbleTestingSignature.cold;
+declare const expectObservable: typeof marbleTestingSignature.expectObservable;
+declare const expectSubscriptions: typeof marbleTestingSignature.expectSubscriptions;
+
+/** @test {smooth} */
```
Comment (staltz) [inline spec/operators/smooth-spec.ts]: I don't think we need this comment.

**Comment by staltz** [comment]
LGTM except for one detail.

File: spec/operators/smooth-spec.ts (hunk 1 above)
Comment (jayphelps) [inline spec/operators/smooth-spec.ts]: hmm the [other tests have them](https://github.com/ReactiveX/rxjs/blob/master/spec/operators/do-spec.ts#L14). Not sure what it's for. figured it was some sort of build time thing.

**Comment by Boshen** [comment]
image

File: spec/operators/smooth-spec.ts (hunk 1 above)
Comment (jayphelps) [inline spec/operators/smooth-spec.ts]: Let's just say it's critical.

**Comment by SteveALee** [comment]
Only for use in diamond shaped flows! :)

**Comment by nvartolomei** [comment]
Does not work in latest Safari :[

**Comment by martinmcwhorter** [comment]
I think there may be an inconsistency in the bridge where it states "coast to coast, LA to Chicago"... Which technically is in a coast... Oh nevermind...

**Comment by saneyuki** [comment]
I love software's easter egg and april fool joke. But why this operator is [released as a part of default bundled operator](https://github.com/ReactiveX/rxjs/commit/e14fa8eb9d4798e6d687100861c782cbe9d41cb1)?

This is not a code comment which can removed by some minifier and imported by default. If user uses `import { Observable } from 'rxjs'` because this operator is loaded by `import './add/operator/smooth';` in `src/Rx.ts`.

This increases the file size of application code which uses rxjs by joke code. I strongly against to ship this operator with `import './add/operator/smooth';` in `src/Rx.ts`.

**Comment by jayphelps** [comment]
Because it was only released in a custom April fools build. `v5.2.1-smooth`

**Comment by saneyuki** [comment]
@jayphelps 

> Because it was only released in a custom April fools build. v5.2.1-smooth. 😝

It is difficult to know that it's just a joke since it has been published in stable channel and has not been disabled yet in April 2nd (PST).  :tired_face:

**Comment by jayphelps** [comment]
@saneyuki Perhaps you can clarify? It wasn't published as `5.2.1` it was published as `5.2.1-smooth`. Someone would have to knowingly install that, and if they don't realize it's a joke there's not much harm done.

It's going to be removed from master shortly, if that's what you mean--but just cause code is in master doesn't make it stable. 😄

**Comment by pelotom** [comment]
`rxjs@5.2.1-smooth` currently shows up as the latest version in NPM, i.e. it's what you get by default when you `npm install rxjs`. I trust the joke doesn't introduce breaking changes, but it's not very professional to release joke code that gets marked as your latest stable release.

**Comment by martinmcwhorter** [comment]
Come on lads, it's just a bit of craic. You would have had to explicitly
chosen to install the specific version and added the operator. It was a
joke, a few people got it and laughed. That is the beauty of open-source,
there is room for a wink and a nod. My only complaint is there was no
reference to Yacht Rock. Keep it smooth.

**Comment by pelotom** [comment]
> You would have had to explicitly chosen to install the specific version

@martinmcwhorter no, you don't have to explicitly choose that version to get it. If you type `npm install rxjs` you will get it. Again, it's not a huge deal, but it's not great either.

**Comment by pelotom** [comment]
I see latest has now been bumped to `5.3.0`, thanks! 👍

**Pull request merged**